src/IECoreScene/SceneCache.cpp

```diff
@@ -728,8 +697,232 @@ class SceneCache::ReaderImplementation : public SceneCache::Implementation
 			return reader;
 		}
 
+		ConstPathMatcherDataPtr readSet( const Name &name ) const override
+		{
+			auto it = m_cachedSets.find( name );
+
+			if( it != m_cachedSets.end() )
+			{
+				return it->second;
+			}
+
+			ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+			nc = nc->getRoot();
+			nc->convertTagsToSets();
+
+			IECore::PathMatcherDataPtr pathMatcherData = readLocalSet( name );
+
+			NameList childSetNames = readChildSetNames();
+
+			if( std::find( childSetNames.begin(), childSetNames.end(), name ) != childSetNames.end() )
+			{
+				NameList children;
+				childNames( children );
+
+				for( const auto &c : children )
+				{
+					ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+					ConstPathMatcherDataPtr childSets = nc->child( c, SceneInterface::NullIfMissing )->readSet( name );
+					pathMatcherData->writable().addPaths( childSets->readable(), {c} );
+				}
+			}
+
+			m_cachedSets[name] = pathMatcherData;
+			return pathMatcherData;
+		}
+
+		NameList setNames() const override
+		{
+			ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+			nc = nc->getRoot();
+			nc->convertTagsToSets();
+
+			IECore::CompoundDataBasePtr compoundDataBase = getSetCompound();
```

Comment:
Wait, here we're loading the contents of all the sets, just to retrieve the names?

Reply:
> Wait, here we're loading the contents of all the sets, just to retrieve the names?
 Write

Yup - luckily this should be easy to fix.